### PR TITLE
Support 0-size ndarray and fix possible error in __del__ at fft

### LINKED ISSUE #3241: `Invalid number of FFT data points` is not always raised
`cupy.fft.fft(cupy.array([]))` should raise `ValueError` even if `n` is default.

<details>

```
>>> np.fft.fft(np.array([]))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<__array_function__ internals>", line 5, in fft
  File "/path/to/python3.8/site-packages/numpy-1.18.2-py3.8-linux-x86_64.egg/numpy/fft/_pocketfft.py", line 188, in fft
    output = _raw_fft(a, n, axis, False, True, inv_norm)
  File "/path/to/python3.8/site-packages/numpy-1.18.2-py3.8-linux-x86_64.egg/numpy/fft/_pocketfft.py", line 57, in _raw_fft
    raise ValueError("Invalid number of FFT data points (%d) specified."
ValueError: Invalid number of FFT data points (0) specified.
>>> cupy.fft.fft(cupy.array([]))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/path/to/cupy/cupy/fft/fft.py", line 496, in fft
    return _fft(a, (n,), (axis,), norm, cupy.cuda.cufft.CUFFT_FORWARD)
  File "/path/to/cupy/cupy/fft/fft.py", line 175, in _fft
    a = _fft_c2c(a, direction, norm, axes, overwrite_x, plan=plan)
  File "/path/to/cupy/cupy/fft/fft.py", line 142, in _fft_c2c
    a = _exec_fft(a, direction, 'C2C', norm, axis, overwrite_x, plan=plan)
  File "/path/to/cupy/cupy/fft/fft.py", line 91, in _exec_fft
    batch = a.size // a.shape[-1]
ZeroDivisionError: integer division or modulo by zero
>>> cupy.fft.fft(cupy.array([]), n=0)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/path/to/cupy/cupy/fft/fft.py", line 496, in fft
    return _fft(a, (n,), (axis,), norm, cupy.cuda.cufft.CUFFT_FORWARD)
  File "/path/to/cupy/cupy/fft/fft.py", line 155, in _fft
    raise ValueError(
ValueError: Invalid number of FFT data points (0) specified.
```

### PR DESCRIPTION
**Pull request by okuta**
Fix #3241

### CODE REVIEW

File: cupy/cuda/cufft.pyx [hunk 1]
```diff
@@ -252,26 +252,39 @@ class Plan1d(object):
         cdef Handle plan
         cdef bint use_multi_gpus = 0 if devices is None else 1
 
+        self.plan = 0
```
[the remark this answers is not in the document]
Comment (okuta) [reply cupy/cuda/cufft.pyx:255]: Thanks, this is redundant.

**Comment by takagi** [comment]
pfnCI, test this please.

File: cupy/cuda/cufft.pyx (hunk 1 above)
[the remark this answers is not in the document]
Comment (takagi) [reply cupy/cuda/cufft.pyx:255]: There is `self.plan = plan` below?

File: cupy/fft/fft.py [hunk 2]
```diff
@@ -493,6 +501,12 @@ def _fftn(a, s, axes, norm, direction, value_type='C2C', order='A', plan=None,
     # Note: need to call _cook_shape prior to sorting the axes
     a = _cook_shape(a, s, axes, value_type, order=order)
 
+    print(a.shape, s, axes_sorted)
```
Comment (takagi) [inline cupy/fft/fft.py:504]:
```suggestion
```

File: cupy/cuda/cufft.pyx (hunk 1 above)
[the remark this answers is not in the document]
Comment (leofang) [reply cupy/cuda/cufft.pyx:255]: I meant `self.plan` will always be nonzero (a valid handle), unless we encounter OOM when allocating the work area and making the plan. In that case, we error out, so it makes little sense to initialize `self.plan` to 0 (the `Plan*d` object won't be usable anyway). Also, note that the `Plan*d` objects are plain Python classes, not cdef classes, so their attributes are dynamic.

Moreover, if the invalid input (batch = 0) is detected in `cupy/fft/fft.py`, we will already raise an error there, so it make further less sense to touch `cupy/cuda/cufft.pyx`, as the code path `if batch == 0` there will never be executed.

[the remark this answers is not in the document]
Comment (okuta) [reply cupy/cuda/cufft.pyx:255]: If you catch an exception that occurred during initialization, `__del__` will also reoccur.
```
class Foo:
    def __init__(self):
        assert False

    def __del__(self):
        print(self.piyo)

try:
    a = Foo()
except:
    pass
```

```
% python foo.py
Exception ignored in: <function Foo.__del__ at 0x7fa28df7fc20>
Traceback (most recent call last):
  File "foo.py", line 6, in __del__
    print(self.piyo)
AttributeError: 'Foo' object has no attribute 'piyo'
```

File: cupy/cuda/cufft.pyx [hunk 3]
```diff
@@ -713,23 +722,28 @@ class PlanNd(object):
             result = cufftCreate(&plan)
             if result == 0:
                 result = cufftSetAutoAllocation(plan, 0)
-            if result == 0:
-                result = cufftMakePlanMany(plan, ndim, shape_ptr,
-                                           inembed_ptr, istride, idist,
-                                           onembed_ptr, ostride, odist,
-                                           <Type>fft_type, batch,
-                                           &work_size)
+        check_result(result)
```
Comment (leofang) [inline cupy/cuda/cufft.pyx:725]: You should hold the plan immediately after it's created. Remember to remove Line 758 below that does the same thing.
```suggestion
        check_result(result)
        self.plan = plan
```

Comment (okuta) [reply cupy/cuda/cufft.pyx:725]: Thanks, I fixed!

**Comment by okuta** [comment]
pfnCI, test this please.

**Comment by leofang** [comment]
Oops, `cupy/statistics/histogram.py` is changed 😂

**Comment by okuta** [comment]
Oh...

**Comment by okuta** [comment]
jenkins, test this please.

**Comment by takagi** [comment]
LGTM!

**Comment by takagi** [comment]
It looks much diverged to make a backport automatically.

**Pull request merged**